src/evaluators/ForStatement.js

```diff
@@ -159,10 +160,36 @@ function ForBodyEvaluation(
 
       // ii. Perform ? GetValue(incRef).
       Environment.GetValue(realm, incRef);
+    } else if (realm.useAbstractInterpretation) {
+      // If we have no increment and we've hit 100 iterations of trying to evaluate
+      // this loop body, then see if we have a break, return or throw completion in a
+      // guarded condition and fail if it does.
+      possibleInfiniteLoopIterations++;
+      if (possibleInfiniteLoopIterations > 100) {
```

Comment:
There is already a timeout check to guard against loops that are actually infinite. In theory that also works in this case. The reason why we need this is because keeping track of so many forked execution paths is highly non linear and can lead to so much work being done from one timeout check to the next that it seems like things will never terminate. It would be nice to have a comment explaining this in the actual source code.